openassessment/xblock/static/js/src/lms/oa_response.js

```diff
@@ -474,71 +537,259 @@ OpenAssessment.ResponseView.prototype = {
+                this.baseView.toggleActionError(
+                    'upload',
+                    gettext("File type is not allowed.")
+                );
+                errorCheckerTriggered = true;
+                break;
+            }
+        }
+
+        if (!errorCheckerTriggered) {
             this.baseView.toggleActionError('upload', null);
             this.files = files;
+            this.updateFilesDescriptionsFields(files, descriptions, uploadType);
+        }
+
+        if (this.files === null) {
+            sel.find('.file__upload').prop('disabled', true);
+        }
+    },
+
+    /**
+     Render textarea fields to input description for each uploaded file.
+
+     */
+    /* jshint -W083 */
+    updateFilesDescriptionsFields: function(files, descriptions, uploadType) {
+        var filesDescriptions = $(this.element).find('.files__descriptions').first();
+        var mainDiv = null;
+        var divLabel = null;
+        var divTextarea = null;
+        var divImage = null;
+        var img = null;
+        var textarea = null;
+        var descriptionsExists = true;
+
+        this.filesDescriptions = descriptions || [];
+
+        $(filesDescriptions).show().html('');
+
+        for (var i = 0; i < files.length; i++) {
+            mainDiv = $('<div/>');
+
+            divLabel = $('<div/>');
+            divLabel.addClass('submission__file__description__label');
+            divLabel.text(gettext("Describe ") + files[i].name + ' ' + gettext("(required):"));
+            divLabel.appendTo(mainDiv);
+
+            divTextarea = $('<div/>');
+            divTextarea.addClass('submission__file__description');
+            textarea = $('<textarea />');
+            if ((this.filesDescriptions.indexOf(i) !== -1) && (this.filesDescriptions[i] !== '')) {
+                textarea.val(this.filesDescriptions[i]);
+            } else {
+                descriptionsExists = false;
+            }
+            textarea.addClass('file__description file__description__' + i);
+            textarea.appendTo(divTextarea);
+
+            if (uploadType === "image") {
+                img = $('<img/>', {
+                    src: window.URL.createObjectURL(files[i]),
+                    height: 80
+                });
+                img.onload = function() {
+                    window.URL.revokeObjectURL(this.src);
+                };
+
+                divImage = $('<div/>');
+                divImage.addClass('submission__img__preview');
+                img.appendTo(divImage);
+                divImage.appendTo(mainDiv);
+            }
+
+            divTextarea.appendTo(mainDiv);
+
+            mainDiv.appendTo(filesDescriptions);
+            textarea.on("change keyup drop paste", $.proxy(this, "checkFilesDescriptions"));
         }
-        $(".file__upload").prop('disabled', this.files === null);
+
+        $(this.element).find('.file__upload').prop('disabled', !descriptionsExists);
```

Comment:
How hard would it be to disable the `Submit your response` button while the descriptions are being edited? Currently it seems a bit strange to have it be available without doing anything. Could you disable it here, and re-enable in `removeFileDescriptions`?

<img width="860" alt="screen shot 2017-04-26 at 4 06 40 pm" src="https://cloud.githubusercontent.com/assets/7373924/25454480/92eb5ab0-2a9a-11e7-9997-11b006896ea7.png">